### ISSUE TITLE
keeping parameters 0-to-1 in backprop: regularization / what-have-you

### ISSUE THREAD

**Issue by maxsolomonhenry** [turian/ddsp-drum#64]
In terms of how you do ML when many of the parameters *should* be at the extrema of 0 or 1:
1) Pass the "preparameter" through a sigmoid-squashing function to turn it into a synth parameter. This is good IF you are okay with the the parameter getting close to 0 / 1 but never hitting it.
2) Pass the "preparameter" through a RELU-like squashing function to clip it, basically. The downside is you get no gradient (like dead neurons) so I think this is a no-go.
3) Add a steep penalty to the loss function if the parameter goes outside [0 - eps, 1 + eps], eps is maybe like 1e-4. Before you do a forward, the ML clips the parameters to 0, 1, *not the synth stuff*. Then it can get a gradient that is like: "Okay we want this parameter less than 0" from the synth but also know "okay but if I make this parameter further from 0 I get penalized so keep the parameter within eps of 0 and that sweet spot means the parameter is actually 0". This is the most promising approach I've tried.

_Originally posted by @turian in https://github.com/turian/ddsp-drum/pull/17#discussion_r554015001_
